sre-recipes/recipe_runner.py

```diff
@@ -0,0 +1,264 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# -*- coding: utf-8 -*-
+
+"""
+This file contains utility runtime classes implementing core SRE Recipes
+features, such as breaking and restoring microservices, printing hints, and
+running interactive multiple choice questions.
+
+Currently, it implements two SRE Recipe Runner:
+- ImplBasedRecipeRunner: runs SRE Recipe implemented via python classes.
+- ConfigBasedRecipeRunner: runs SRE Recipes defined as YAML configs.
+
+Refer to the class docstring for further explanations.
+"""
+
+import abc
+import importlib
+import requests
+import subprocess
+import yaml
+
+from inspect import isclass
+from os import path
+
+import utils
+from recipes.impl_based.base import BaseRecipeImpl
+
+
+# Default Load Generation Config
+DEFAULT_LOADGEN_USER_TYPE = "BasicHomePageViewingUser"
+DEFAULT_LOADGEN_USER_COUNT = 20
+DEFAULT_LOADGEN_SPAWN_RATE = 1
+DEFAULT_LOADGEN_TIMEOUT_SECONDS = 600
+
+
+class ImplBasedRecipeRunner:
+    """A SRE Recipe runner for running recipes implemented as class objects.
+
+    Given a `recipe_name`, it tries to run `recipes/impl_based/recipe_name.py`.
+
+    This runner will propgate all exceptions to the caller, and it is caller's
+    responsibility to handle any exception and to perform any error logging.
+    """
+
+    def __init__(self, recipe_name):
+        self.recipe = None
+        module = importlib.import_module(f"recipes.impl_based.{recipe_name}")
+        for attribute_name in dir(module):
+            attr = getattr(module, attribute_name)
+            if isclass(attr) and attr is not BaseRecipeImpl and issubclass(attr, BaseRecipeImpl):
+                self.recipe = attr()
+                break
+        if not self.recipe:
+            raise NotImplementedError(
+                f"No valid implementation exists for `{recipe_name}` recipe.")
+
+    def get_name(self):
+        return self.recipe.get_name()
+
+    def get_description(self):
+        return self.recipe.get_description()
+
+    def run_break(self):
+        return self.recipe.run_break()
+
+    def run_restore(self):
+        return self.recipe.run_restore()
+
+    def run_hint(self):
+        return self.recipe.run_hint()
+
+    def run_verify(self):
+        return self.recipe.run_verify()
+
+
+class ConfigBasedRecipeRunner:
+    """A SRE Recipe runner for running recipes implemented using configs.
+
+    Given a `recipe_name`, it tries to load `recipes/configs_based/recipe_name.yaml`.
+
+    This runner will propgate all exceptions to the caller, and it is caller's
+    responsibility to handle any exception and to perform any error logging.
+    """
+
+    def __init__(self, recipe_name):
+        filepath = path.join(path.dirname(
+            path.abspath(__file__)), f"recipes/configs_based/{recipe_name}.yaml")
+        with open(filepath, "r") as file:
+            self.recipe = yaml.safe_load(file.read())
```

Comment:
We cannot. But since the parsed YAML is in JSON format, we can use JSON-Schema library to do validation checks. I am planning to do that as a part of the follow-up PR for integration testing and validations.

https://stackoverflow.com/questions/3262569/validating-a-yaml-document-in-python